File: margot_heel/margot_heel_cli/margot_heel_cli/model_monitor.py
from . import model_parameter
import sys




class MonitorModel:
  """
  This class represents a monitor, in particular it defines:
    - the monitor class
    - the monitor header
    - the monitor type
    - the monitor name
    - the monitor stop method name
    - the monitor start method name
    - the list of creation parameters
    - the list of [required] start parameters
    - the list of [required] stop parameters
  """


  def __init__(self):
    # The available output metrics
    self.output_available_metrics = ['AVERAGE', 'STDDEV', 'MAX', 'MIN']


    # The monitor specs
    self.monitor_name = 'my_default_monitor'
    self.monitor_class = 'margot::Monitor<float>'
    self.monitor_header = '<margot/monitor.hpp>'
    self.monitor_type = 'float'
    self.stop_method = ''
    self.start_method = ''
    self.required_start_methods = [] # the name of the required parameters
    self.required_creation_methods = []
    self.required_stop_methods = []

    # The input parameters
    self.creation_parameters = [] # the list of the parsed parameters
    self.start_parameters = []
    self.stop_parameters = []

    # The output parameters key -> what, value -> var name
    self.exposed_metrics = {}

    # Error monitor specific parameters
    self.frequency = ''
    self.period = ''


  def check_consistency(self):
    """
    This method checks if all the parameters required are defined
    """
    for pr in self.required_start_methods:
      found = False
      for p in self.start_parameters:
        if p.param_name == pr.param_name:
          found = True
          break
      if not found:
        print('[CONSISTENCY ERROR] The "{0}" monitor with name "{1}" has an incomplete definition:'.format(self.monitor_class, self.monitor_name))
        print('                    Unable to to find the required input parameter "{0}" '.format(pr.param_name))
        print('                    Defined start parameters -> ["{0}"]'.format('", "'.join([a.param_name for a in self.start_parameters])))
        sys.exit(-1)




  def __str__(self):
    """
    dump the monitor model
    """
    dump_string = '\n'

    dump_string = '{0}\n  Monitor specification'.format(dump_string)
    dump_string = '{0}\n    Name:   {1}'.format(dump_string, self.monitor_name)
    dump_string = '{0}\n    Class:  {1}'.format(dump_string, self.monitor_class)
    dump_string = '{0}\n    Header: {1}'.format(dump_string, self.monitor_header)
    dump_string = '{0}\n    Type:   {1}'.format(dump_string, self.monitor_type)

    dump_string = '{0}\n  Monitor usage (pseudocode)'.format(dump_string)
    dump_string = '{0}\n    Constructor:  {1}({2})'.format(dump_string, self.monitor_class, ','.join([p.param_name for p in self.creation_parameters]))
    if self.start_method:
      dump_string = '{0}\n    Start method: {1}({2})'.format(dump_string, self.start_method, ','.join([p.param_name for p in self.start_parameters]))
    else:
      dump_string = '{0}\n    Start method: Not defined'.format(dump_string)

    if self.stop_method:
      dump_string = '{0}\n    Stop method:  {1}({2})'.format(dump_string, self.stop_method, ','.join([p.param_name for p in self.stop_parameters]))
    else:
      dump_string = '{0}\n    Stop method:  Not defined'.format(dump_string)

    dump_string = '{0}\n  Exposed variables'.format(dump_string)
    for exposed_var in self.exposed_metrics:
      dump_string = '{0}\n    "{1}" in a variable with name "{2}"'.format(dump_string, exposed_var.upper(), self.exposed_metrics[exposed_var])

    if self.frequency:
      dump_string = '{0}\n    Error monitor frequency: {1}'.format(dump_string, self.frequency)
    if self.period:
      dump_string = '{0}\n    Error monitor period: {1}'.format(dump_string, self.period)

    return dump_string







know_monitor_spec = [
  'CUSTOM',
  'FREQUENCY',
  'MEMORY',
  'PAPI',
  'CPUPROCESS',
  'CPUSYSTEM',
  'TEMPERATURE',
  'THROUGHPUT',
  'TIME',
  'ENERGY',
  'ODROID_POWER',
  'ODROID_ENERGY',
  'COLLECTOR',
  'ERROR'
]

available_frequencies = ['never', 'always', 'periodic', 'auto']


def get_monitor_spec( monitor_type ):
  """
  Provides the list of known spec
  """

  if monitor_type.upper() == 'FREQUENCY':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::FrequencyMonitor'
    my_monitor.monitor_header = '<margot/frequency_monitor.hpp>'
    my_monitor.monitor_type = 'margot::FrequencyMonitor::value_type'
    my_monitor.stop_method = 'measure'
    my_monitor.start_method = ''

    return my_monitor

  if monitor_type.upper() == 'MEMORY':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::MemoryMonitor'
    my_monitor.monitor_header = '<margot/memory_monitor.hpp>'
    my_monitor.monitor_type = 'margot::MemoryMonitor::value_type'
    my_monitor.stop_method = 'extractMemoryUsage'
    my_monitor.start_method = ''

    return my_monitor

  if monitor_type.upper() == 'PAPI':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::PapiMonitor'
    my_monitor.monitor_header = '<margot/papi_monitor.hpp>'
    my_monitor.monitor_type = 'margot::PapiMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'
    my_monitor.required_creation_methods = ['event']

    return my_monitor

  if monitor_type.upper() == 'CPUPROCESS':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::ProcessCpuMonitor'
    my_monitor.monitor_header = '<margot/process_cpu_usage_monitor.hpp>'
    my_monitor.monitor_type = 'margot::ProcessCpuMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'CPUSYSTEM':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::SystemCpuMonitor'
    my_monitor.monitor_header = '<margot/system_cpu_usage_monitor.hpp>'
    my_monitor.monitor_type = 'margot::SystemCpuMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'TEMPERATURE':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::TemperatureMonitor'
    my_monitor.monitor_header = '<margot/temperature_monitor.hpp>'
    my_monitor.monitor_type = 'margot::TemperatureMonitor::value_type'
    my_monitor.stop_method = 'measure'
    my_monitor.start_method = ''

    return my_monitor

  if monitor_type.upper() == 'THROUGHPUT':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::ThroughputMonitor'
    my_monitor.monitor_header = '<margot/throughput_monitor.hpp>'
    my_monitor.monitor_type = 'margot::ThroughputMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'
    my_monitor.required_stop_methods = ['num_data_elaborated']

    return my_monitor

  if monitor_type.upper() == 'TIME':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::TimeMonitor'
    my_monitor.monitor_header = '<margot/time_monitor.hpp>'
    my_monitor.monitor_type = 'margot::TimeMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'ENERGY':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::EnergyMonitor'
    my_monitor.monitor_header = '<margot/energy_monitor.hpp>'
    my_monitor.monitor_type = 'margot::EnergyMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'COLLECTOR':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::CollectorMonitor'
    my_monitor.monitor_header = '<margot/collector_monitor.hpp>'
    my_monitor.monitor_type = 'margot::CollectorMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'ODROID_POWER':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::OdroidPowerMonitor'
    my_monitor.monitor_header = '<margot/odroid_power_monitor.hpp>'
    my_monitor.monitor_type = 'margot::OdroidPowerMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'ODROID_ENERGY':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::OdroidEnergyMonitor'
    my_monitor.monitor_header = '<margot/odroid_energy_monitor.hpp>'
    my_monitor.monitor_type = 'margot::OdroidEnergyMonitor::value_type'
    my_monitor.stop_method = 'stop'
    my_monitor.start_method = 'start'

    return my_monitor

  if monitor_type.upper() == 'ERROR':
    # create the model
    my_monitor = MonitorModel()

    # set the spec
    my_monitor.monitor_class = 'margot::ErrorMonitor'
    my_monitor.monitor_header = '<margot/error_monitor.hpp>'
    my_monitor.stop_method = 'stop'


    return my_monitor


  # no known spec found
  return None

File: heel/src/model_monitor.cpp
/* mARGOt HEEL library
 * Copyright (C) 2018 Politecnico di Milano
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 */

#include <algorithm>
#include <map>
#include <stdexcept>
#include <vector>

#include <boost/algorithm/string.hpp>

#include <heel/logger.hpp>
#include <heel/model_monitor.hpp>
#include <heel/typer.hpp>

namespace margot {
namespace heel {

// helper function that generates a spec for a custom monitor
inline monitor_spec gen_custom_spec(const std::string& monitor_type) {
  // get the type of a monitor
  const std::string real_type = sanitize_type(monitor_type);
  if (real_type.compare("string") == 0) {
    error("Unable to generate a monitor of strings");
    throw std::runtime_error("custom monitor: unsupported type");
  }
  return {"margot::Monitor<" + real_type + ">",
          "margot/monitor.hpp",
          real_type,
          "",
          "push",
          {{parameter_types::IMMEDIATE, "1", "size_t"}},
          {},
          {{parameter_types::VARIABLE, "element", monitor_type}}};
}

// these are the required information for each type monitor that we have implemented in margot
// NOTE: the actual value type of each monitor should be taken directly from its class... However, since
//       we can't avoid a manual update of this map, it makes no sense to bring mARGOt as dependency just
//       for this reason, it will raise too much the complexity of the building system
static std::map<std::string, monitor_spec> known_monitors = {
    {"collector",
     {"margot::CollectorMonitor",
      "margot/collector_monitor.hpp",
      typer<double>::get(),
      "start",
      "stop",
      {{parameter_types::VARIABLE, "topic", "std::string"},
       {parameter_types::VARIABLE, "address", "std::string"},
       {parameter_types::VARIABLE, "port", "int"},
       {parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"energy",
     {"margot::EnergyMonitor",
      "margot/energy_monitor.hpp",
      typer<long double>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "margot::EnergyMonitor::Domain::Cores", "margot::EnergyMonitor::Domain"},
       {parameter_types::IMMEDIATE, "{}", "std::vector<std::size_t>"},
       {parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"frequency",
     {"margot::FrequencyMonitor",
      "margot/frequency_monitor.hpp",
      typer<unsigned int>::get(),
      "",
      "measure",
      {{parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"memory",
     {"margot::MemoryMonitor",
      "margot/memory_monitor.hpp",
      typer<std::size_t>::get(),
      "",
      "extractMemoryUsage",
      {{parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"odroid_energy",
     {"margot::OdroidEnergyMonitor",
      "margot/odroid_energy_monitor.hpp",
      typer<long double>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "margot::TimeUnit::MICROSECONDS", "margot::TimeUnit"},
       {parameter_types::VARIABLE, "polling_time_ms", "uint64_t"},
       {parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"odroid_power",
     {"margot::OdroidPowerMonitor",
      "margot/odroid_power_monitor.hpp",
      typer<float>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"papi",
     {"margot::PapiMonitor",
      "margot/papi_monitor.hpp",
      typer<long long int>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "margot::PapiEvent::L1_MISS", "margot::PapiEvent"},
       {parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"process_cpu",
     {"margot::ProcessCpuMonitor",
      "margot/process_cpu_usage_monitor.hpp",
      typer<float>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "margot::CounterType::SoftwareCounter", "margot::CounterType"},
       {parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"system_cpu",
     {"margot::SystemCpuMonitor",
      "margot/system_cpu_usage_monitor.hpp",
      typer<float>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"temperature",
     {"margot::TemperatureMonitor",
      "margot/temperature_monitor.hpp",
      typer<long long int>::get(),
      "",
      "measure",
      {{parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}},
    {"throughput",
     {"margot::ThroughputMonitor",
      "margot/throughput_monitor.hpp",
      typer<float>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {{parameter_types::VARIABLE, "amount_data", "float"}}}},
    {"time",
     {"margot::TimeMonitor",
      "margot/time_monitor.hpp",
      typer<unsigned long int>::get(),
      "start",
      "stop",
      {{parameter_types::IMMEDIATE, "margot::TimeUnit::MICROSECONDS", "margot::TimeUnit"},
       {parameter_types::IMMEDIATE, "1", "size_t"}},
      {},
      {}}}};

static const std::vector<std::string> available_statistics = {"average", "standard_deviation", "max", "min"};

const monitor_spec& get_monitor_cpp_spec(monitor_model& monitor) {
  // make sure that we support the given monitor type
  boost::trim(monitor.type);
  try {
    return get_monitor_cpp_spec(monitor.type);
  } catch (const std::runtime_error& e) {
    error("Unknown type \"", monitor.type, "\" for monitor \"", monitor.name, "\"");
    throw e;
  }
}

const monitor_spec& get_monitor_cpp_spec(const std::string& monitor_type) {
  auto monitor_spec_it = known_monitors.find(monitor_type);
  if (monitor_spec_it == known_monitors.cend()) {
    // if reach this statement, is not a known monitor, but we could generate the spec on the fly if it is a
    // numeric type (i.e. custom monitor)
    const auto res = known_monitors.emplace(monitor_type, gen_custom_spec(monitor_type));
    monitor_spec_it = res.first;
  }
  return monitor_spec_it->second;
}

void validate(monitor_model& model) {
  // check if the monitor has a type
  if (model.type.empty()) {
    error("The monitor \"", model.name, "\" must have a type");
    throw std::runtime_error("monitor model: monitor without a type");
  }

  // now we need to consider all the parameters in the monitor and combine them with the ones in the
  // knowledge. The idea is that the user should either redefine the whole set, or use the default version
  // we define this lambda to apply this procedure to all the parameters. Moreover, we set as a prefix the
  // name of the monitor, to make the variable name unique
  const auto check_parameters = [&model](const std::vector<struct parameter>& default_params,
                                         std::vector<struct parameter>& params,
                                         const std::string& params_name) {
    if (params.empty()) {
      params = default_params;
    } else if (params.size() != default_params.size()) {
      error("Unable to partially define the ", params_name, " parameters of monitor \"", model.name,
            "\". Found ", params.size(), " parameter(s), ", default_params.size(), " expected");
      throw std::runtime_error("monitor model: mismatch on " + params_name + " parameters");
    }
    std::transform(params.begin(), params.end(), params.begin(), [&model](parameter param) {
      if (param.type == parameter_types::VARIABLE) {
        param.content = model.name + "_" + param.content;
      }
      return param;
    });
  };

  // now we can check all the parameters of the monitor
  const auto spec = get_monitor_cpp_spec(model);
  check_parameters(spec.default_param_initialization, model.initialization_parameters, "initialization");
  check_parameters(spec.default_param_start, model.start_parameters, "start");
  check_parameters(spec.default_param_stop, model.stop_parameters, "stop");

  // now we need to check if the requested statistic to log is ok
  std::for_each(model.requested_statistics.begin(), model.requested_statistics.end(),
                [&model](const std::string& statistic_name) {
                  if (!std::any_of(available_statistics.begin(), available_statistics.end(),
                                   [&statistic_name](const std::string& known_name) {
                                     return known_name.compare(statistic_name) == 0;
                                   })) {
                    error("Unable to compute the statistic \"", statistic_name, "\" in monitor \"",
                          model.name, "\"");
                    throw std::runtime_error("Unknown log statistic");
                  }
                });

  // check if the name is a valid c/c++ identifier
  if (!is_valid_identifier(model.name)) {
    error("The monitor name \"", model.name, "\" is not a valid c/c++ identifier");
    throw std::runtime_error("monitor model: unsupported name");
  }
}

bool is_custom_monitor(const monitor_model& monitor) {
  return known_monitors.at(monitor.type).stop_method_name.compare("push") == 0;
}

}  // namespace heel
}  // namespace margot